src/tiered/tiered_cursor.c

```diff
@@ -1046,12 +1033,26 @@ __curtiered_next_random(WT_CURSOR *cursor)
     __cursor_novalue(cursor);
     WT_ERR(__curtiered_enter(curtiered, false));
 
-    for (;;) {
-        __curtiered_random_tier(session, curtiered, &c);
-        /*
-         * This call to next_random on the tier can potentially end in WT_NOTFOUND if the tier we
-         * picked is empty. We want to retry in that case.
-         */
+    /*
+     * Select a random item from a random tier. Keep trying different tiers until we either find
+     * something or run out of tiers. To do this we generate a random permutation of the tier
+     * cursors and iterate through them in the resulting order.
+     *
+     * TODO: We should weight this by the number of elements in each tier.
```

Comment:
I'm cognizant of keeping the number of TODOs in the code low.  If we expect this to be handled soon, by us, I would use the convention `TODO: tiered: ...` .  Or we could file a ticket.

But I think the bigger issue is that each "next" call does a memory allocation, which seems rather heavyweight.  Perhaps just selecting a random tier and if it is NOT_FOUND, go to the next one in order (wrapping), and quit if all are exhausted?  

However we do it, I think the TODO comment is right that we may need weighting.  But the number of elements in a btree is not something we maintain directly, though there's probably surrogate numbers (number of leaf pages?)

Reply:
> I'm cognizant of keeping the number of TODOs in the code low. If we expect this to be handled soon, by us, I would use the convention `TODO: tiered: ...` . Or we could file a ticket.

I think there are a bunch of TODO items (including this one) that were left in the original prototype code to indicate things Michael didn't get around to implementing. I'll move it to a ticket because I don't think we need to come up with something better right now. And by waiting, we'll have a better idea of where and how tiered cursors will be used and what other statistics might be available (or desirable) to improve the randomness.

> But I think the bigger issue is that each "next" call does a memory allocation, which seems rather heavyweight. Perhaps just selecting a random tier and if it is NOT_FOUND, go to the next one in order (wrapping), and quit if all are exhausted?

TBH, I wasn't too worried about the memory allocation since random cursor operations are pretty rare and often slow to start with (e.g., see WT-7373).  But I like the suggestion of just picking a random tier and then iterating in order, since it's simpler and probably just as effective.  So I've implemented that.

Reply:
Also incorporated @ddanderson's suggestion for simplifying the error path.